### PR TITLE
Describe all loops in the sequence diagram

### PR DESCRIPTION
**Pull request by Anya497**
- Add description of running _Hyperparameters tuning_ and _Dataset improvement with SVMs_ 
- Clarify documentation for ONNX converter

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,47 +43,85 @@ Then follow installation instructions provided on [torch](https://pytorch.org/ge
...
+_Illustration of game process:_
+<div align="center">
+<img src="resources/game_process_illustration.png" width="50%" title="Illustration of game process"/>
+</div>
+
+As shown in the picture there are two different loops in the _Main Loop_. To run _Hyper-parameters Tuning Loop_ do steps listed in [this chapter](#hyper-parameters-tuning). The second loop runs symbolic execution with given methods and updates the dataset. To run it do [these steps](#dataset-improvement-with-symbolic-virtual-machines).
+The _Main Loop_ implies alternation of these two variants of running as much as you want. After the _Main Loop_ you can run symbolic execution with the trained model (checkout [this chapter](#guide-symbolic-execution-with-trained-model)).
+
+### Hyper-parameters tuning
+
```
Comment (gsvgit) [inline README.md:61]: Brief intro required.

File: README.md [hunk 2]
```diff
@@ -43,47 +43,85 @@ Then follow installation instructions provided on [torch](https://pytorch.org/ge
...
+    Now initial dataset saved in the directory `./AIAgent/report/SerializedEpisodes`. Then it will be updated by neural network if it finds a better solution.
+2) Create configuration (specifying training parameters). You can use [`./workflow/config_for_tests.yml`](./workflow/config_for_tests.yml) as a template.
+  To use the loss function that is used for training as objective function for hyper-parameters tuning, set the validation config as follow:
+    ```yml
+    ValidationConfig:
+      validation:
+        val_type: loss
+        batch_size: <DEPENDS_ON_YOUR_RAM_SIZE>
+    ```
+1) Move to **AIAgent** directory
```
Comment (gsvgit) [inline README.md:75]: Enumeration reset. Intro missed?

File: README.md [hunk 3]
```diff
@@ -43,47 +43,85 @@ Then follow installation instructions provided on [torch](https://pytorch.org/ge
...
-- Create configuration (specifying server and training parameters). You can use [`./workflow/config_for_tests.yml`](./workflow/config_for_tests.yml) as a template.
-  - Below we will discuss how you can fully integrate your own symbolic machine
-- Launch the server manager
-  - `poetry run python3 launch_servers.py --config path/to/config.yml`
-- Run MLFlow server
-  - `poetry run mlflow server -h 127.0.0.1 -p 8080 --serve-artifacts`
-- Run the training process to get PyTorch model.
-  - `poetry run python3 run_training.py --config path/to/config.yml`
+### Guide symbolic execution with trained model
+1) Use [`onyx.py`](#onnx-conversion) command line to tool convert your PyTorch model to ONNX format.
```
Comment (gsvgit) [inline README.md:115]: `command line to tool convert` --- misprint?

File: README.md [hunk 4]
```diff
@@ -43,47 +43,85 @@ Then follow installation instructions provided on [torch](https://pytorch.org/ge
...
+        batch_size: <DEPENDS_ON_YOUR_RAM_SIZE>
+    ```
+1) Move to **AIAgent** directory
+    ```sh
+    cd AIAgent
+    ```
+2) Run the training process.
+    ```sh
+    poetry run python3 run_training.py --config path/to/config.yml
+    ```
```
[the remark this answers is not in the document]
Comment (Anya497) [reply README.md:82]: 1) It's used in both cases and it consists of only one command. I think that one command in the separate chapter is not beautiful.
2) Yes, it doesn't need SVMs. So it works without server launch.

File: README.md [hunk 5]
```diff
@@ -43,47 +43,85 @@ Then follow installation instructions provided on [torch](https://pytorch.org/ge
...
+
+_Illustration of game process:_
+<div align="center">
+<img src="resources/game_process_illustration.png" width="50%" title="Illustration of game process"/>
+</div>
+
+As shown in the picture there are two different loops in the _Main Loop_. To run _Hyper-parameters Tuning Loop_ do steps listed in [this chapter](#hyper-parameters-tuning). The second loop runs symbolic execution with given methods and updates the dataset. To run it do [these steps](#dataset-improvement-with-symbolic-virtual-machines).
+The _Main Loop_ implies alternation of these two variants of running as much as you want. After the _Main Loop_ you can run symbolic execution with the trained model (checkout [this chapter](#guide-symbolic-execution-with-trained-model)).
+
+### Hyper-parameters tuning
```
[the remark this answers is not in the document]
Comment (Anya497) [reply README.md:60]: 1) Yes, it's better to present some example.
2) I can't understand why it is less complete. Could you write more about it, please?
3) Hm. I think it's better to create separate chapter for dataset generation because we need to do it once, not every time when run hyper-parameters tuning. 

But the main pipeline consists of hyper-parameters tuning and `Dataset improvement with Symbolic Virtual Machines` as presented in the sequence diagram.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (emnigma) [reply README.md:93]: > 2. Yes, it doesn't need SVMs. So it works without server launch.

Then we should specify that we have two different modes: `Hyper-parameters tuning` and `Dataset improvement`. Why should we run tuning before dataset improvement?

File: README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (emnigma) [reply README.md:67]: > 1. Yes, it's better to present some example.

👍 

> 2. I can't understand why it is less complete. Could you write more about it, please?

I was under the impression that hyperparameter tuning happens inside the main pipeline and requires server launch

> 3. Hm. I think it's better to create separate chapter for dataset generation because we need to do it once, not every time when run hyper-parameters tuning.

👍 
 
> But the main pipeline consists of hyper-parameters tuning and `Dataset improvement with Symbolic Virtual Machines` as presented in the sequence diagram.

👍

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Anya497) [reply README.md:93]: > > 2. Yes, it doesn't need SVMs. So it works without server launch.
> 
> Then we should specify that we have two different modes: `Hyper-parameters tuning` and `Dataset improvement`. Why should we run tuning before dataset improvement?

I think it is specified in [this](https://github.com/PySymGym/PySymGym/pull/170/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R57) paragraph.

**Review by gsvgit — APPROVED** [review]

**Review by emnigma — APPROVED** [review]
Great job! Also before the paper comes out, we should spellcheck this

**Review by Parzival-05 — APPROVED** [review]
👍

File: README.md [hunk 6]
```diff
@@ -43,46 +43,103 @@ Then follow installation instructions provided on [torch](https://pytorch.org/ge
 
 ## Usage
 
-### Follow these steps:
+We use [**MLFlow**](https://mlflow.org/) to log hyper-parameters and training artifacts. Firstly, run MLFlow server:
+```sh
+cd AIAgent
```
[the remark this answers is not in the document]
Comment (Anya497) [reply README.md:48]: MLFlow creates special directories for logging artifacts inside the directory were it was launched from and in my opinion placing them inside `AIAgent` is better.
